providers/base/bin/prime_offload_tester.py

```diff
@@ -0,0 +1,334 @@
+# along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+import threading
+import subprocess
+import time
+import re
+import json
+import argparse
+import logging
+from enum import IntEnum
+
+
+class PrimeOffloaderError(IntEnum):
+    """
+    A class used to define PrimeOffloader Error code
+
+    :attr NO_ERROR: process success
+    :type NO_ERROR: int
+
+    :attr NO_CARD_ID: couldn't find card id
+    :type NO_CARD_ID: int
+
+    :attr NO_CARD_NAME: couldn't find card name
+    :type NO_CARD_NAME: int
+
+    :attr OFFLOAD_FAIL: couldn't find process on specific GPU
+    :type OFFLOAD_FAIL: int
+
+    :attr NOT_SUPPORT_NV_PRIME: this system couldn't support nv prime offload
+    :type NOT_SUPPORT_NV_PRIME: int
+
+    :attr NV_DRIVER_ERROR: some errors form nvidia driver
+    :type NV_DRIVER_ERROR: int
+
+    """
+
+    NO_ERROR = 0
+    NO_CARD_ID = -1
+    NO_CARD_NAME = -2
+    OFFLOAD_FAIL = -3
+    NOT_SUPPORT_NV_PRIME = -4
+    NV_DRIVER_ERROR = -5
+
+
+class PrimeOffloader:
+    """
+    A class used to execute process to specific GPU.
+    Have to run this as root.
+
+    :attr logger: console logger
+    :type logger: obj
+
+    :attr check_result:
+        store the result of checking offloading is ok or not.
+    :type check_result: int
+    """
+
+    logger = logging.getLogger()
+
+    check_result = PrimeOffloaderError.OFFLOAD_FAIL
+
+    def find_card_id(self, pci_name):
+        """
+        use pci name to find card id under /sys/kernel/debug/dri
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type card_name: str
+        """
+        cmd = "sudo grep -lr --include=name \"{}\""  \
+              " /sys/kernel/debug/dri 2>/dev/null".format(pci_name)
+        try:
+            if not re.match("[0-9]{4}:[0-9]{2}:[0-9]{2}.[0-9]", pci_name):
+                return PrimeOffloaderError.NO_CARD_ID
+
+            card_path = subprocess.run(cmd, shell=True,
+                                       stdout=subprocess.PIPE,
+                                       universal_newlines=True)
+            card_id = card_path.stdout.split('/')[5]
+            return card_id
+        except (ValueError, IndexError, AttributeError):
+            return PrimeOffloaderError.NO_CARD_ID
+
+    def find_card_name(self, pci_name):
+        """
+        use pci name to find card name by lshw
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type card_name: str
+        """
+        cmd = 'sudo lshw -c display -json'
+        try:
+            card_infos = subprocess.run(cmd, shell=True,
+                                        stdout=subprocess.PIPE,
+                                        universal_newlines=True)
+            infos = json.loads(card_infos.stdout)
+            for info in infos:
+                if pci_name in info['businfo']:
+                    return info['product']
+        except (ValueError, IndexError, TypeError):
+            return None
```

Comment:
This makes the return type become a union (str|None).
Is not obtaining the name a normal case or an exceptional case? If it's an exception, maybe it's worth making it as such.